runtime/lua/vim/_watch.lua

```diff
@@ -72,129 +76,127 @@ function M.watch(path, opts, callback)
+  local function is_excluded(filepath)
+    return opts.exclude_pattern and opts.exclude_pattern:match(filepath)
   end
-  if not watches.is_dir and not incl_match() or excl_match() then
-    return watches.cancel
+
+  local process_changes = function()
+    assert(false, "Replaced later. I'm only here as forward reference")
   end
 
-  if not watches.handle then
-    local poll, new_err = vim.uv.new_fs_poll()
-    assert(not new_err, new_err)
-    watches.handle = poll
-    local _, start_err = poll:start(
-      path,
-      interval,
-      vim.schedule_wrap(function(err)
-        if err == 'ENOENT' then
-          return
-        end
-        assert(not err, err)
-        poll_internal(path, opts, callback, watches)
-        callback(path, M.FileChangeType.Changed)
-      end)
-    )
-    assert(not start_err, start_err)
-    if watches.started then
-      callback(path, M.FileChangeType.Created)
+  local function create_on_change(filepath)
+    return function(err, filename, events)
+      assert(not err, err)
+      local fullpath = vim.fs.joinpath(filepath, filename)
+      if is_included(fullpath) and not is_excluded(filepath) then
+        table.insert(changesets, {
+          fullpath = fullpath,
+          events = events,
+        })
+        timer:stop()
+        timer:start(debounce, 0, process_changes)
+      end
     end
   end
 
-  if watches.is_dir then
-    watches.children = watches.children or {}
-    local exists = {} --- @type table<string,true>
-    for name, ftype in vim.fs.dir(path) do
-      exists[name] = true
-      if not watches.children[name] then
-        watches.children[name] = {
-          is_dir = ftype == 'directory',
-          started = watches.started,
-        }
-        poll_internal(filepath_join(path, name), opts, callback, watches.children[name])
+  process_changes = function()
+    ---@type table<string, table[]>
+    local filechanges = vim.defaulttable()
+    for i, change in ipairs(changesets) do
+      changesets[i] = nil
+      if is_included(change.fullpath) and not is_excluded(change.fullpath) then
+        table.insert(filechanges[change.fullpath], change.events)
       end
     end
-
-    local newchildren = {} ---@type table<string,watch.Watches>
-    for name, watch in pairs(watches.children) do
-      if exists[name] then
-        newchildren[name] = watch
+    for fullpath, events_list in pairs(filechanges) do
+      local stat = uv.fs_stat(fullpath)
+      ---@type vim._watch.FileChangeType
+      local change_type
+      if stat then
+        change_type = FileChangeType.Created
+        for _, event in ipairs(events_list) do
+          if event.change then
+            change_type = FileChangeType.Changed
+          end
+        end
+        if stat.type == 'directory' then
+          local handle = assert(uv.new_fs_event())
+          handles[fullpath] = handle
+          handle:start(fullpath, uvflags, create_on_change(fullpath))
```

Comment:
It's not clear to me why we need to create a new watcher on changes to a directory? Can this result in multiple watchers on the same directory?

Reply:
If a new directory gets created within a watched directory we need to start watching it for changes too. The stat call above should ensure it didn't exist,  but I'll add a check to see if a handle already exists in case there's some weird create-delete-create chain

Reply:
The stat call on line 136? Doesn't that just ensure the path exists?

Reply:
Ah wait. I think I follow since the watcher only checks for deletes, creates and renamed.

Reply:
Yes, my thinking was that a folder either gets created or deleted, and otherwise there is no event.  But I actually didn't check what happens on permission changes or a touch.

In any case,  I pushed a fixup to check if a handle already exists